### PR TITLE
RFC: required propTypes no longer generate null and undefined

### PR DESCRIPTION
**Pull request by mxstbr**
Previously, even prop types like `PropTypes.string.isRequired` generated `null` and `undefined` on random generation, which they shouldn't.

This is a RFC because I'm not sure the signature for the `randomValues` function makes sense. I changed it to:

```JS
Control.randomValue = (props) { /* … */ }
/* …or, for Array and Object controls… */
Control.randomValue = (propTypeValue) { /* … */ }
```

This `props`/`propTypeValue` argument has an `required` property, which is injected in `utils/getControls.js` for each component.
Every control is aware of the fact that it can or cannot generate null or undefined. This should be overrideable in the admin interface, which is why I still re-determine it on every `randomValue()` call.

Maybe it makes more sense to change the function signature to `randomValue = (someObject)`, which would allow us to pass more advanced options in there too. (like min/max for numbers, number type,…)

What do you think?

### CODE REVIEW

File: plugin/default-plugins/playground/frontend/components/controls/ObjectControl/index.js [hunk 1]
```diff
@@ -49,14 +51,14 @@ const ObjectControl = ({ label, propTypeData, value, onUpdate, isNested }) => {
...
-  const canBeUndefined = true;
+  const canReallyBeNull = !propTypeData.required;
+  const canReallyBeUndefined = !propTypeData.required;
   const normalizedPropsWithControls = mapValues(propTypeData.value, (prop) => {
     prop.control = getControl(prop); // eslint-disable-line no-param-reassign
     return prop;
   });
   const value = randomValues(normalizedPropsWithControls);
-  return valueOrNullOrUndefined(value, canBeNull, canBeUndefined);
+  return valueOrNullOrUndefined(value, canReallyBeNull, canReallyBeUndefined);
```
Comment (nikgraf) [inline plugin/default-plugins/playground/frontend/components/controls/ObjectControl/index.js]: hahaha, I suggest we pick `canReallyReallyReallyBeNull` which on my humble opinion sound stronger than `canReallyBeNull `

File: plugin/default-plugins/playground/frontend/components/controls/ArrayControl/index.js [hunk 2]
```diff
@@ -57,17 +57,23 @@ const ArrayControl = (props) => {
...
   const max = 4;
   const size = Math.floor(Math.random() * (max - min + 1)) + min;
   const rangeArray = range(min, size);
-  const propTypeData = props.value || props.type && props.type.value; // TODO clean up
-  const control = getControl(propTypeData);
+  // Get the prop type data of the insides of the array
+  const innerPropTypeData =
+    propTypeData.value
+    || propTypeData.type
+    && propTypeData.type.value; // TODO clean up
```
Comment (nikgraf) [inline plugin/default-plugins/playground/frontend/components/controls/ArrayControl/index.js]: Offtopic: I wonder if we can remove `type.value` here already? Aren't we normalizing it already?

**Comment by nikgraf** [comment]
Handling `required` was something I had in mind for quite some time, but didn't tackle and this here is a good solution especially that it's eventually overridable in the admin interface.

regarding `randomValue = (someObject)`: makes sense, in the old signature this information could have been in the random property of the props. I'm not entirely sure how I had it in mind or why we did it like this.

Wanna wait there? I think we will run into it in a few hours when we integrate the server and get min/max info as well.
Feel free to merge the PR.

File: plugin/default-plugins/playground/frontend/components/controls/ObjectControl/index.js (hunk 1 above)
Comment (mxstbr) [inline plugin/default-plugins/playground/frontend/components/controls/ObjectControl/index.js]: Oops, that's an artifact from an earlier idea, this should simply be `canBeNull` and `canBeUndefined`. Will fix, good catch!

File: plugin/default-plugins/playground/frontend/components/controls/ArrayControl/index.js (hunk 2 above)
Comment (mxstbr) [inline plugin/default-plugins/playground/frontend/components/controls/ArrayControl/index.js]: No idea, I'm not very confident in that part of the code anymore…

**Pull request merged**